slsim/Sources/Supernovae/supernovae_pop.py

```diff
@@ -0,0 +1,105 @@
+import numpy as np
+import scipy.integrate as integrate
+import scipy.interpolate as interp
+
+"""References:
+Oguri and Marshall 2010
+"""
+
+
+def calculate_star_formation_rate(z):
+    """Calculates the cosmic star formation rate. (Eq 13 - Oguri and Marshall 2010)
+
+    :param z: redshift
+    :return: cosmic star formation rate in [(h)yr^(-1)Mpc^(-3)]
+    """
+    if z >= 0:
+        star_formation_rate = (0.0118 + 0.08 * z) / (1 + (z / 3.3) ** 5.2)
+        return star_formation_rate
+    else:
+        return np.nan
+
+
+def t_d_power(t_d):
+    """Calculates the power law constraint on time delay. (Eq 14 - Oguri and Marshall 2010)
+
+    :param t_d: time delay
+    :return: constrained time delay
+    """
+    if t_d > 0:
+        ft_d = (t_d) ** (-1.08)
+        return ft_d
+    else:
+        return np.nan
+
+
+class SNIaRate(object):
+    """Class to calculate supernovae rates."""
+
+    def __init__(self, cosmo):
+        """
+        :param cosmo: cosmology object
+        :type cosmo: ~astropy.cosmology object
+        """
+        self._cosmo = cosmo
+
+    def z_from_time(self, t):
+        """Calculates redshift given cosmic time.
+
+        :param t: cosmic time since big bang in [Gy]
+        :param cosmo: cosmology object
+        :type cosmo: ~astropy.cosmology object
+        :return: redshift at time t [float]
+        """
+        if not hasattr(self, "_age_inv"):
+            z_array = np.linspace(0, 1000, 100000)
```

Comment:
you might only go to z=100 (fyi, the highest redshift galaxy is at z~14)

Reply:
do you need 100k interpolation points to be accurate enough? Just asking if we can afford to be a bit less accurate for speed

Reply:
I used 100k interpolation points to get more accuracy closer to 0 for the sake of the plots but later added a point to the array well below 0 for this purpose, so I will lower both numbers.